### ISSUE TITLE
S2583 False Positive for expression containing null conditional and null coalesce

### ISSUE THREAD

**Issue by alex-grigoras** [SonarSource/sonar-dotnet#4168]
### Description

Expressions like `var a = x?.func() ?? false;` are marked as not compliant with rule S2583.
SonarQube reports that the expression always evaluates to `false` when the value of the expression is actually dependant on the value returned by `func`.

### Repro steps

I wrote a small unit test for this issue. Drop it in ExplodedGraphTests.cs.
```
[TestMethod]
[TestCategory("Symbolic execution")]
public void ExplodedGraph_NullConditional_And_Coalesce()
{
    const string testInput = "var a = x?.func() ?? false;";
    var context = new ExplodedGraphContext(testInput);
    var aSymbol = context.GetSymbol("a");
    context.ExplodedGraph.InstructionProcessed +=
        (sender, args) =>
        {
            if (args.Instruction.ToString() == "a = x?.func() ?? false")
            {
                args.ProgramState.GetSymbolValue(aSymbol).Should().NotBe(SymbolicValue.True);
                args.ProgramState.GetSymbolValue(aSymbol).Should().NotBe(SymbolicValue.False);
            }
        };

    context.WalkWithInstructions(8);
}
```

### Expected behavior

SonarQube should not trigger S2583 for expressions like `var a = x?.func() ?? false;` because they are not always false.

### Actual behavior

SonarQube gives that error.
At first I thought it was an issue related to https://github.com/SonarSource/sonar-dotnet/issues/2369 which was fixed in https://github.com/SonarSource/sonar-dotnet/pull/2487 but I debugged the CFG code and it looked ok, at least to my n00b eyes.
I suspect an issue with the Symbolic Execution. I debugged it to the code in CSharpExplodedGraph.cs and saw that when the value for the symbol `a` gets popped of the stack it is `SV_False` when it should in fact be the value of `func()` if `x` is not null. 

### Related information

* SonarQube Community EditionVersion 7.9.1 (build 27448)
* seen on both .net core 3.1 and .net framework 4.7.2

**Comment by agrigoras-bd** [comment]
Symbolic executing looks OK, there's a bug in my test. That event is raised more than once, with the value of "a" being "SV_False" once and "SV_5" the other time, which is as expected.

**Comment by agrigoras-bd** [comment]
The test case below does not trigger the rule, I tried it with the unit test ConditionEvaluatesToConstant_FromCSharp8.

`
using System;
using System.Collections.Generic;

namespace Tests.Diagnostics
{
    public class DummyClass
    {
        private static Random random = new Random();

        bool Dummy()
        {
            return random.NextDouble() == 0;
        }

        DummyClass XFactory()
        {
            return random.NextDouble() == 0? new DummyClass() : null;
        }

        bool ShouldBeGood()
        {
            var x = XFactory();
            return x?.Dummy() ?? false;
        }
    }
}
`

**Comment by agrigoras-bd** [comment]
You can close this issue as not reproducible, I don't know where it's from and cannot get it to happen again.

**Comment by csaba-sagi-sonarsource** [comment]
Hi @agrigoras-bd, 

As I could not reproduce this issue with the snippet you sent, I will close the issue as you suggested.
If you manage to create a reproducer for the issue, feel free to reopen the issue.
